// staking.2.0/lib/buttonBlock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ButtonBlock = function ButtonBlock(props) {
  var stakeMoreDisabled = props.stakeMoreDisabled,
      stakeDisabled = props.stakeDisabled,
      nextDisabled = props.nextDisabled;
  var styleBtnGreen = {
    background: '#0BFFB7',
    padding: 12,
    width: 180,
    cursor: 'pointer',
    border: 'none',
    textTransform: "uppercase",
    marginBlock: 20,
    marginInline: 10,
    fontSize: 12
  };
  var styleBtnRed = {
    background: '#FB5252',
    padding: 12,
    width: 180,
    cursor: 'pointer',
    border: 'none',
    textTransform: "uppercase",
    marginBlock: 20,
    marginInline: 10,
    color: "#fff",
    fontSize: 12
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      textAlign: 'center'
    },
    className: "button-block-style"
  }, props.onClickStakeMore && /*#__PURE__*/_react.default.createElement("button", {
    onClick: props.onClickStakeMore,
    style: styleBtnGreen,
    className: "button-block-style-btn-green"
  }, "Stake More"), props.onClickStake && /*#__PURE__*/_react.default.createElement("button", {
    disabled: stakeDisabled,
    onClick: props.onClickStake,
    style: styleBtnGreen,
    className: "button-block-style-btn-green"
  }, "Stake"), props.onClickNext && !nextDisabled && /*#__PURE__*/_react.default.createElement("button", {
    onClick: props.onClickNext,
    style: styleBtnGreen,
    className: "button-block-style-btn-green"
  }, props.text), props.onClickNext && nextDisabled && /*#__PURE__*/_react.default.createElement("button", {
    disabled: true,
    onClick: props.onClickNext,
    style: styleBtnGreen,
    className: "button-block-style-btn-green"
  }, "Next"), props.onClickRequest && /*#__PURE__*/_react.default.createElement("button", {
    onClick: props.onClickRequest,
    style: styleBtnRed,
    className: "button-block-style-btn-red"
  }, "REQUEST WITHDRAW"), props.onClickWithdrawal && /*#__PURE__*/_react.default.createElement("button", {
    onClick: props.onClickWithdrawal,
    style: styleBtnRed,
    className: "button-block-style-btn-red"
  }, "WITHDRAW"));
};

var _default = ButtonBlock;
exports.default = _default;
