// staking.2.0/lib/infoBlock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ContentCopy = _interopRequireDefault(require("@mui/icons-material/ContentCopy"));

var _Info = _interopRequireDefault(require("@mui/icons-material/Info"));

var _IconButton = _interopRequireDefault(require("@mui/material/IconButton"));

var _Tooltip = _interopRequireDefault(require("@mui/material/Tooltip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BORDER_COLOR = 'rgba(255, 255, 255, 0.18)';
var GRAY_COLOR = 'rgba(255, 255, 255, 0.50)';
var content = {
  alignItems: 'center',
  marginTop: 20
};
var addressStyle = {
  color: '#fff',
  fontSize: 13,
  marginHorizontal: 30,
  textAlign: 'center',
  marginTop: 10
};
var row = {
  marginTop: 10,
  //   marginInline : 20,
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'center',
  alignItems: 'center',
  paddingVertical: 10,
  borderTopWidth: 0.5,
  borderTopColor: BORDER_COLOR,
  borderTopStyle: 'solid',
  borderBottomWidth: 0.5,
  borderBottomColor: BORDER_COLOR,
  borderBottomStyle: 'solid'
};
var column = {
  flex: 0.5,
  maxHeight: 'auto',
  justifyContent: 'center',
  alignItems: 'center',
  display: 'flex',
  flexDirection: "column",
  marginBlock: 12,
  position: 'relative'
};
var column2 = {
  flex: 0.5,
  maxHeight: 'auto',
  justifyContent: 'center',
  alignItems: 'center',
  borderLeftWidth: 0.5,
  borderLeftColor: BORDER_COLOR,
  borderLeftStyle: 'solid',
  display: 'flex',
  flexDirection: "column",
  marginBlock: 12,
  position: 'relative'
};
var value = {
  color: '#fff',
  fontSize: 14,
  textAlign: 'center'
};
var subtitle = {
  color: GRAY_COLOR,
  fontSize: 11,
  textTransform: 'uppercase',
  marginTop: 10
};
var link_active_stake = 'https://support.velas.com/hc/en-150/articles/360021044820-Delegation-Warmup-and-Cooldown';

var Info = function Info(props) {
  return /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
    title: /*#__PURE__*/_react.default.createElement("div", null, props.titleInfo, " ", props.link && /*#__PURE__*/_react.default.createElement("a", {
      href: link_active_stake,
      target: "_blank",
      style: {
        color: '#0BFFB7',
        cursor: 'pointer',
        textDecoration: 'underline'
      }
    }, "Read More.")),
    style: {
      position: 'absolute',
      top: -10,
      right: 5
    }
  }, /*#__PURE__*/_react.default.createElement(_IconButton.default, null, /*#__PURE__*/_react.default.createElement(_Info.default, {
    color: "inherit",
    style: {
      color: "#FFF",
      fontSize: 10
    }
  })));
};

var InfoBlock = function InfoBlock(props) {
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
    style: content,
    className: "info-block-content"
  }, props.name && /*#__PURE__*/_react.default.createElement("div", {
    style: addressStyle,
    className: "info-block-address-style",
    onClick: props.copyName
  }, props.name), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      color: props.name ? 'rgba(255, 255, 255, 0.3)' : '#fff',
      fontSize: 12,
      marginHorizontal: 30,
      marginBlock: 20,
      textAlign: 'center',
      marginTop: 10,
      cursor: "pointer"
    },
    className: "info-block-address",
    onClick: props.copyAddress
  }, props.address, " ", /*#__PURE__*/_react.default.createElement(_ContentCopy.default, {
    fontSize: "15"
  }))), /*#__PURE__*/_react.default.createElement("div", {
    style: row,
    className: "info-block-row"
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: column,
    className: "info-block-column"
  }, props.titleInfo && /*#__PURE__*/_react.default.createElement(Info, props), /*#__PURE__*/_react.default.createElement("div", {
    style: value,
    className: "info-block-value"
  }, props.value2), /*#__PURE__*/_react.default.createElement("div", {
    style: subtitle,
    className: "info-block-subtitle"
  }, props.subtitle2)), /*#__PURE__*/_react.default.createElement("div", {
    style: column2,
    className: "info-block-column2"
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: value,
    className: "info-block-value"
  }, props.value1), /*#__PURE__*/_react.default.createElement("div", {
    style: subtitle,
    className: "info-block-subtitle"
  }, props.subtitle1))));
};

var _default = InfoBlock;
exports.default = _default;
