// staking.2.0/lib/loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _CircularProgress = _interopRequireDefault(require("@mui/material/CircularProgress"));

var _Box = _interopRequireDefault(require("@mui/material/Box"));

var _styleSpinner;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var styleSpinner = (_styleSpinner = {
  top: '48%',
  position: "absolute",
  left: '48%',
  color: "#fff"
}, _defineProperty(_styleSpinner, "left", 0), _defineProperty(_styleSpinner, "right", 0), _defineProperty(_styleSpinner, "textAlign", 'center'), _styleSpinner);
var containerStyle = {
  position: 'absolute',
  zIndex: 999,
  top: 0,
  bottom: 0,
  left: 0,
  right: 0
};
var msgStyle = {
  marginBottom: 20
};

var Loader = function Loader(props) {
  var text = props.text,
      show = props.show;
  if (!show) return null;
  return /*#__PURE__*/_react.default.createElement("div", {
    class: "ss",
    style: containerStyle
  }, /*#__PURE__*/_react.default.createElement(_Box.default, {
    sx: styleSpinner
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: msgStyle
  }, text || ""), /*#__PURE__*/_react.default.createElement(_CircularProgress.default, {
    color: "inherit"
  })));
};

var _default = Loader;
exports.default = _default;
