// staking.2.0/lib/stakeItem.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactJdenticon = _interopRequireDefault(require("react-jdenticon"));

var _badge = _interopRequireDefault(require("./badge"));

var _bn = _interopRequireDefault(require("bn.js"));

var _formatValue = require("./format-value");

var _List = _interopRequireDefault(require("@mui/material/List"));

var _ListItemAvatar = _interopRequireDefault(require("@mui/material/ListItemAvatar"));

var _ListItemButton = _interopRequireDefault(require("@mui/material/ListItemButton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// import index from './index.css'
var StakeItem = function StakeItem(props) {
  var address = props.address;
  var addressCut = address.substring(0, 15) + '...' + address.substring(35);
  var isStaked = props.myStake > 0 ? true : false;
  return /*#__PURE__*/React.createElement(_List.default, {
    sx: {
      width: '100%',
      cursor: "pointer"
    },
    style: props.style
  }, /*#__PURE__*/React.createElement(_ListItemButton.default, {
    onClick: props.onClick
  }, /*#__PURE__*/React.createElement(_ListItemAvatar.default, null, /*#__PURE__*/React.createElement(_reactJdenticon.default, {
    size: "40",
    value: props.address
  })), /*#__PURE__*/React.createElement("div", {
    style: {
      display: 'flex',
      justifyContent: 'center',
      width: '100%',
      fontSize: 12
    }
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      alignItems: 'flex-start',
      width: '100%',
      marginBlock: 5
    }
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      marginTop: 2,
      color: props.name ? 'rgba(255,255,255,0.50)' : '#fff',
      fontSize: 12
    }
  }, addressCut), props.name && /*#__PURE__*/React.createElement("div", {
    style: {
      marginBlock: 2,
      fontSize: 12
    }
  }, props.name), /*#__PURE__*/React.createElement("div", {
    style: {
      marginTop: 6,
      color: 'rgba(255,255,255,0.50)',
      fontSize: 10
    }
  }, isStaked ? "My Stake" : "Total Staked"), /*#__PURE__*/React.createElement("div", {
    style: {
      marginBlock: 2,
      color: isStaked ? '#0BFFB7' : "#fff"
    }
  }, " ", isStaked ? (0, _formatValue.formatStakeAmount)(props.myStake) : (0, _formatValue.formatStakeAmount)(props.totalStaked), ' ', "VLX")), /*#__PURE__*/React.createElement("div", {
    style: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      alignItems: 'flex-end',
      width: '100%',
      marginBlock: 5
    }
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      marginTop: 6
    }
  }, /*#__PURE__*/React.createElement(_badge.default, {
    status: props.status
  })), /*#__PURE__*/React.createElement("div", {
    style: {
      height: 10
    }
  }), /*#__PURE__*/React.createElement("div", {
    style: {
      marginTop: 6,
      color: 'rgba(255,255,255,0.50)',
      fontSize: 10
    }
  }, "APR,%"), /*#__PURE__*/React.createElement("div", {
    style: {
      marginBlock: 2
    }
  }, null !== props.apr && (props.apr * 100).toFixed(2))))));
};

var _default = StakeItem;
exports.default = _default;
