// staking.2.0/lib/topRowTable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styleHeader = {
  backgroundColor: "#04051B",
  color: "#fff",
  fontSize: 13
};

var TopRow = function TopRow(props) {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      backgroundColor: "#04051B",
      height: 40,
      display: "flex",
      fontSize: 12
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 80,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.first), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 240,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.second), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 240,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.third, ",%"), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 110,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.fourth, " (VLX)"), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 50,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.fifth, ",%"), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 100,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.sixth), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 100,
      textAlign: 'center',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    }
  }, props.seventh));
};

var _default = TopRow;
exports.default = _default;
