diff --git a/node_modules/@velas/web3/lib/index.cjs.js b/node_modules/@velas/web3/lib/index.cjs.js
index 6e54c56..c93327d 100644
--- a/node_modules/@velas/web3/lib/index.cjs.js
+++ b/node_modules/@velas/web3/lib/index.cjs.js
@@ -10,7 +10,7 @@ var bs58 = require('bs58');
 var cryptoHash = require('crypto-hash');
 var borsh = require('borsh');
 var BufferLayout = require('@solana/buffer-layout');
-var fetch = require('node-fetch');
+var fetch = require('cross-fetch');
 var superstruct = require('superstruct');
 var rpcWebsockets = require('rpc-websockets');
 var RpcClient = require('jayson/lib/client/browser');
