// staking.2.0/lib/images/enterance-img.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var EnteranceImg = function EnteranceImg(props) {
  return /*#__PURE__*/React.createElement("svg", {
    width: props.width,
    height: props.height,
    viewBox: "0 0 322 345",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M145.223 15.6801C149.381 14.3318 151.657 9.8759 150.305 5.72762C148.954 1.57934 144.487 -0.690493 140.329 0.657807C136.17 2.00611 133.895 6.46197 135.247 10.6102C136.598 14.7585 141.065 17.0284 145.223 15.6801Z",
    fill: "#FFA607"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M12.5823 113.026C15.4099 112.109 16.9571 109.079 16.038 106.258C15.119 103.437 12.0817 101.894 9.25412 102.811C6.42651 103.727 4.87931 106.757 5.79836 109.578C6.71741 112.399 9.75468 113.943 12.5823 113.026Z",
    fill: "#302E75"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M289.931 259.274C292.533 256.902 292.714 252.875 290.336 250.279C287.959 247.683 283.922 247.502 281.32 249.874C278.718 252.246 278.536 256.273 280.914 258.869C283.292 261.465 287.329 261.646 289.931 259.274Z",
    fill: "#FFA607"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M313.786 104.509C315.852 102.626 315.996 99.4289 314.108 97.3681C312.22 95.3072 309.015 95.1633 306.95 97.0466C304.884 98.9298 304.74 102.127 306.627 104.188C308.515 106.249 311.72 106.393 313.786 104.509Z",
    fill: "#FFA607"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M111.211 284.111C113.277 282.228 113.421 279.03 111.533 276.97C109.645 274.909 106.44 274.765 104.374 276.648C102.309 278.531 102.164 281.729 104.052 283.79C105.94 285.85 109.145 285.994 111.211 284.111Z",
    fill: "#4442A6"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M276.61 151.007C276.61 194.635 249.969 232.28 211.766 248.485C198.948 253.97 184.622 256.961 169.793 256.961C154.964 256.961 140.638 253.97 127.82 248.485C89.869 232.28 62.9763 194.635 62.9763 151.007C62.9763 107.379 89.869 69.7339 127.82 53.5291C140.638 48.0444 154.964 45.0527 169.793 45.0527C184.622 45.0527 198.948 48.0444 211.766 53.5291C249.969 69.7339 276.61 107.379 276.61 151.007Z",
    fill: "url(#paint0_linear_7746_24817-995258)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M205.481 48.0449C207.491 50.467 209.502 52.6468 211.764 55.0688C232.625 75.4139 246.197 110.049 246.197 149.77C246.197 189.249 232.373 224.127 211.764 244.472C208.999 247.136 206.486 250.043 204.224 252.949",
    stroke: "#2F2D74",
    strokeWidth: "6.43328",
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M135.612 252.949C133.35 249.778 130.585 246.851 127.82 244.167C106.959 223.921 93.3875 188.795 93.3875 148.789C93.3875 109.028 107.211 73.9019 127.82 53.4115C129.58 51.7039 131.339 49.9964 132.847 48.0449",
    stroke: "#2F2D74",
    strokeWidth: "6.43328",
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M169.794 45.0527L169.793 257.46",
    stroke: "#2F2D74",
    strokeWidth: "6.43328",
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M268.569 110.866H71.0208",
    stroke: "#2F2D74",
    strokeWidth: "6.43328",
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M268.569 191.646H71.0208",
    stroke: "#2F2D74",
    strokeWidth: "6.43328",
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M282.352 151.008C282.352 213.173 231.957 263.568 169.792 263.568C107.626 263.568 57.2312 213.173 57.2312 151.008C57.2312 88.8423 107.626 38.4473 169.792 38.4473C231.957 38.4473 282.352 88.8423 282.352 151.008ZM64.473 151.008C64.473 209.174 111.626 256.326 169.792 256.326C227.957 256.326 275.11 209.174 275.11 151.008C275.11 92.8418 227.957 45.6891 169.792 45.6891C111.626 45.6891 64.473 92.8418 64.473 151.008Z",
    fill: "#C4C4C4"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M282.352 151.008C282.352 213.173 231.957 263.568 169.792 263.568C107.626 263.568 57.2312 213.173 57.2312 151.008C57.2312 88.8423 107.626 38.4473 169.792 38.4473C231.957 38.4473 282.352 88.8423 282.352 151.008ZM64.473 151.008C64.473 209.174 111.626 256.326 169.792 256.326C227.957 256.326 275.11 209.174 275.11 151.008C275.11 92.8418 227.957 45.6891 169.792 45.6891C111.626 45.6891 64.473 92.8418 64.473 151.008Z",
    fill: "url(#paint1_linear_7746_24817-258816)"
  }), /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0_7746_24817-335575)"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M101.324 257.715C134.189 257.715 160.832 231.072 160.832 198.207C160.832 165.341 134.189 138.699 101.324 138.699C68.4586 138.699 41.8159 165.341 41.8159 198.207C41.8159 231.072 68.4586 257.715 101.324 257.715Z",
    fill: "#EAB14D"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.324 138.702C100.261 138.702 99.2058 138.731 98.1572 138.786C129.55 140.432 154.499 166.407 154.499 198.21C154.499 230.013 129.55 255.988 98.1572 257.634C99.2058 257.689 100.261 257.718 101.324 257.718C134.189 257.718 160.832 231.075 160.832 198.21C160.832 165.344 134.189 138.702 101.324 138.702Z",
    fill: "#E49542"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.324 244.582C126.934 244.582 147.696 223.82 147.696 198.21C147.696 172.6 126.934 151.839 101.324 151.839C75.7141 151.839 54.9529 172.6 54.9529 198.21C54.9529 223.82 75.7141 244.582 101.324 244.582Z",
    fill: "#F4FBFF"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.325 151.839C100.041 151.839 98.7698 151.893 97.512 151.996C121.338 153.935 140.07 173.884 140.07 198.21C140.07 222.536 121.338 242.485 97.512 244.425C98.7698 244.527 100.041 244.582 101.325 244.582C126.936 244.582 147.697 223.821 147.697 198.21C147.697 172.6 126.936 151.839 101.325 151.839Z",
    fill: "#DAF1F4"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.322 166.114C100.33 166.114 99.5261 165.31 99.5261 164.318V161.682C99.5261 160.69 100.33 159.886 101.322 159.886C102.314 159.886 103.118 160.69 103.118 161.682V164.318C103.118 165.31 102.314 166.114 101.322 166.114Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M125.193 176.08C124.733 176.08 124.274 175.905 123.923 175.554C123.222 174.853 123.222 173.716 123.923 173.014L125.787 171.151C126.489 170.449 127.626 170.449 128.327 171.151C129.029 171.852 129.029 172.989 128.327 173.691L126.463 175.555C126.112 175.905 125.653 176.08 125.193 176.08Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M137.659 200.009H135.023C134.031 200.009 133.227 199.205 133.227 198.213C133.227 197.221 134.031 196.417 135.023 196.417H137.659C138.651 196.417 139.455 197.221 139.455 198.213C139.455 199.205 138.651 200.009 137.659 200.009Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M126.92 225.743C126.46 225.743 126 225.568 125.65 225.217L123.786 223.353C123.084 222.652 123.084 221.515 123.786 220.813C124.487 220.112 125.624 220.112 126.326 220.813L128.19 222.677C128.891 223.378 128.891 224.515 128.19 225.217C127.839 225.568 127.379 225.743 126.92 225.743Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.125 236.345C100.133 236.345 99.3291 235.541 99.3291 234.549V231.913C99.3291 230.921 100.133 230.117 101.125 230.117C102.117 230.117 102.921 230.921 102.921 231.913V234.549C102.921 235.541 102.117 236.345 101.125 236.345Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M75.3939 225.603C74.9343 225.603 74.4745 225.428 74.124 225.077C73.4224 224.376 73.4227 223.239 74.124 222.537L75.988 220.673C76.6896 219.972 77.8265 219.972 78.528 220.674C79.2293 221.375 79.2293 222.512 78.5278 223.214L76.6638 225.077C76.3132 225.428 75.8534 225.603 75.3939 225.603Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M67.4276 199.813H64.7915C63.7997 199.813 62.9956 199.009 62.9956 198.017C62.9956 197.025 63.7997 196.221 64.7915 196.221H67.4276C68.4194 196.221 69.2235 197.025 69.2235 198.017C69.2237 199.009 68.4194 199.813 67.4276 199.813Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M77.3969 175.946C76.9373 175.946 76.4777 175.77 76.127 175.42L74.2629 173.556C73.5616 172.854 73.5614 171.717 74.2627 171.016C74.9642 170.315 76.1012 170.314 76.8027 171.016L78.6667 172.88C79.3681 173.581 79.3683 174.718 78.667 175.42C78.316 175.77 77.8564 175.946 77.3969 175.946Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.324 200.008H86.0708C85.079 200.008 84.2749 199.203 84.2749 198.211C84.2749 197.219 85.079 196.415 86.0708 196.415H99.5282V173.566C99.5282 172.574 100.332 171.77 101.324 171.77C102.316 171.77 103.12 172.574 103.12 173.566V198.211C103.12 199.203 102.316 200.008 101.324 200.008Z",
    fill: "#2B4D66"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M119.751 209.483C119.475 209.483 119.194 209.419 118.931 209.284L100.504 199.808C99.6222 199.355 99.2749 198.272 99.7284 197.39C100.182 196.508 101.265 196.16 102.147 196.614L120.574 206.089C121.456 206.543 121.804 207.626 121.35 208.508C121.031 209.127 120.403 209.483 119.751 209.483Z",
    fill: "#DA4A54"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M101.324 203.007C103.971 203.007 106.117 200.861 106.117 198.213C106.117 195.566 103.971 193.419 101.324 193.419C98.6762 193.419 96.53 195.566 96.53 198.213C96.53 200.861 98.6762 203.007 101.324 203.007Z",
    fill: "#365E7D"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear_7746_24817-995258",
    x1: "193.17",
    y1: "96.3159",
    x2: "68.549",
    y2: "254.815",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#25245C"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.364583",
    stopColor: "#2A2968"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#4543A9"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint1_linear_7746_24817-258816",
    x1: "152.738",
    y1: "171.437",
    x2: "284.73",
    y2: "36.6468",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#25245C"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.364583",
    stopColor: "#2A2968"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#4543A9"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0_7746_24817-335575"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "119.016",
    height: "119.016",
    fill: "white",
    transform: "translate(41.8159 138.702)"
  }))));
};

var _default = EnteranceImg;
exports.default = _default;
