// staking.2.0/lib/images/exit-validator-img.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var ExitValidatorImg = function ExitValidatorImg(props) {
  return /*#__PURE__*/React.createElement("svg", {
    width: props.width,
    height: props.height,
    viewBox: "0 0 322 322",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M272.917 160.999C272.917 191.482 260.467 218.959 240.503 238.708C220.968 258.243 193.92 270.05 164.081 270.05C134.457 270.05 107.409 258.028 87.66 238.708C67.696 218.959 55.2454 191.482 55.2454 160.999C55.2454 100.678 103.975 51.9482 164.081 51.9482C224.188 51.9482 272.917 100.892 272.917 160.999Z",
    fill: "#EAEEF9"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M272.917 160.999C272.917 191.482 260.467 218.959 240.503 238.708C220.968 258.243 193.92 270.05 164.081 270.05C134.457 270.05 107.409 258.028 87.66 238.708C67.696 218.959 55.2454 191.482 55.2454 160.999C55.2454 100.678 103.975 51.9482 164.081 51.9482C224.188 51.9482 272.917 100.892 272.917 160.999Z",
    fill: "url(#paint0_linear_7746_24695-452222)"
  }), /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0_7746_24695-583465)"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M197.673 174.51C187.761 174.51 179.695 166.444 179.695 156.532C179.695 146.62 187.761 138.554 197.673 138.554H209.658V114.584C209.658 104.672 201.591 96.6055 191.679 96.6055H127.258C125.286 96.6055 123.441 97.5755 122.32 99.1995C121.205 100.823 120.954 102.892 121.661 104.731L165.108 218.593C165.989 220.918 168.218 222.452 170.706 222.452H191.679C201.591 222.452 209.658 214.386 209.658 204.474V174.51H197.673Z",
    fill: "#5A81E5"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M225.34 186.04C223.105 185.11 221.642 182.923 221.642 180.502V162.523H197.672C194.363 162.523 191.678 159.839 191.678 156.531C191.678 153.223 194.363 150.539 197.672 150.539H221.642V132.561C221.642 130.139 223.105 127.952 225.34 127.023C227.581 126.094 230.158 126.61 231.871 128.324L255.843 152.294C258.186 154.637 258.186 158.425 255.843 160.768L231.871 184.739C230.158 186.452 227.581 186.968 225.34 186.04Z",
    fill: "white"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M165.587 109.198L129.577 97.1951C121.45 94.6966 113.773 100.695 113.773 108.587V216.456C113.773 221.581 117.034 226.153 121.888 227.831L157.891 239.834C159.216 240.241 160.433 240.428 161.716 240.428C168.326 240.428 173.7 235.052 173.7 228.442V120.573C173.7 115.45 170.44 110.877 165.587 109.198Z",
    fill: "#2735B2"
  })), /*#__PURE__*/React.createElement("path", {
    d: "M289.548 185.853C294.409 185.853 298.35 181.912 298.35 177.051C298.35 172.19 294.409 168.25 289.548 168.25C284.688 168.25 280.747 172.19 280.747 177.051C280.747 181.912 284.688 185.853 289.548 185.853Z",
    fill: "#302E75"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M261.008 69.4895C263.22 69.4895 265.012 67.6971 265.012 65.486C265.012 63.2749 263.22 61.4824 261.008 61.4824C258.797 61.4824 257.005 63.2749 257.005 65.486C257.005 67.6971 258.797 69.4895 261.008 69.4895Z",
    fill: "#FFA607"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M58.6796 89.3016C61.9992 89.3016 64.6903 86.6105 64.6903 83.2909C64.6903 79.9713 61.9992 77.2803 58.6796 77.2803C55.36 77.2803 52.6689 79.9713 52.6689 83.2909C52.6689 86.6105 55.36 89.3016 58.6796 89.3016Z",
    fill: "#302E75"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M34.8128 233.211C40.9778 233.211 45.9755 228.213 45.9755 222.048C45.9755 215.883 40.9778 210.886 34.8128 210.886C28.6478 210.886 23.6501 215.883 23.6501 222.048C23.6501 228.213 28.6478 233.211 34.8128 233.211Z",
    fill: "#FFA607"
  }), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear_7746_24695-452222",
    x1: "147.279",
    y1: "180.791",
    x2: "244.152",
    y2: "88.1948",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0.138163",
    stopColor: "#25245C"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#4543A9"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0_7746_24695-583465"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "143.827",
    height: "143.827",
    fill: "white",
    transform: "translate(113.774 96.6006)"
  }))));
};

var _default = ExitValidatorImg;
exports.default = _default;
